src/vs/editor/browser/services/treeSitter/treeSitterParserService.ts

```diff
@@ -54,10 +51,18 @@ export class TextModelTreeSitter extends Disposable {
 	 */
 	private async _onDidChangeLanguage(languageId: string) {
 		this._languageSessionDisposables.clear();
-		this._treeSitterTree = undefined;
+		this._parseResult = undefined;
 
 		const token = cancelOnDispose(this._languageSessionDisposables);
-		const language = await this._treeSitterParser.getLanguage(languageId);
+		let language = this._treeSitterLanguages.getLanguage(languageId);
+		if (!language) {
+			const languageAdded = Event.toPromise(Event.onceIf(this._treeSitterLanguages.onDidAddLanguage, (e) => e.id === languageId));
+			const canceled = new Promise<void>(resolve => token.onCancellationRequested(resolve));
+			const result = await Promise.race([languageAdded, canceled]);
+			if (result) {
+				language = result.language;
+			}
+		}
```

Comment:
I think the pattern here would be `waitForDefinedState(getState: () => T | undefined, onDidChange: Event<unknown>, cancellationToken): Promise<T>`. The implementation would throw on cancellation and wait for `getState` to return a non-undefined value. It would check it initially and whenever the event fires.
`onceIf` can be dangerous, if the event never fires initially.

Reply:
I've changed to a `_getLanguage` method that throws on cancellation. When we're using the cancellation token to indicate that we should stop processing the language change, I do prefer the promise race though for the consistancy of only having one thing to check: has cancellation occurred.